datafusion/physical-expr/src/simplifier/unwrap_cast.rs

```diff
@@ -0,0 +1,621 @@
+// under the License.
+
+//! Unwrap casts in binary comparisons for physical expressions
+//!
+//! This module provides optimization for physical expressions similar to the logical
+//! optimizer's unwrap_cast module. It attempts to remove casts from comparisons to
+//! literals by applying the casts to the literals if possible.
+//!
+//! The optimization improves performance by:
+//! 1. Reducing runtime cast operations on column data
+//! 2. Enabling better predicate pushdown opportunities
+//! 3. Optimizing filter expressions in physical plans
+//!
+//! # Example
+//!
+//! Physical expression: `cast(column as INT64) > INT64(10)`
+//! Optimized to: `column > INT32(10)` (assuming column is INT32)
+
+use std::sync::Arc;
+
+use arrow::datatypes::{DataType, Schema};
+use datafusion_common::{
+    tree_node::{Transformed, TreeNode},
+    Result, ScalarValue,
+};
+use datafusion_expr::Operator;
+use datafusion_expr_common::casts::try_cast_literal_to_type;
+
+use crate::expressions::{lit, BinaryExpr, CastExpr, Literal, TryCastExpr};
+use crate::PhysicalExpr;
+
+/// Attempts to unwrap casts in comparison expressions.
+pub(crate) fn unwrap_cast_in_comparison(
+    expr: Arc<dyn PhysicalExpr>,
+    schema: &Schema,
+) -> Result<Transformed<Arc<dyn PhysicalExpr>>> {
+    expr.transform_down(|e| {
+        if let Some(binary) = e.as_any().downcast_ref::<BinaryExpr>() {
+            if let Some(unwrapped) = try_unwrap_cast_binary(binary, schema)? {
+                return Ok(Transformed::yes(unwrapped));
+            }
+        }
+        Ok(Transformed::no(e))
+    })
+}
+
+/// Try to unwrap casts in binary expressions
+fn try_unwrap_cast_binary(
+    binary: &BinaryExpr,
+    schema: &Schema,
+) -> Result<Option<Arc<dyn PhysicalExpr>>> {
+    // Case 1: cast(left_expr) op literal
+    if let (Some((inner_expr, _cast_type)), Some(literal)) = (
+        extract_cast_info(binary.left()),
+        binary.right().as_any().downcast_ref::<Literal>(),
+    ) {
+        if let Some(unwrapped) = try_unwrap_cast_comparison(
+            Arc::clone(inner_expr),
+            literal.value(),
+            *binary.op(),
+            schema,
+        )? {
+            return Ok(Some(unwrapped));
+        }
+    }
+
+    // Case 2: literal op cast(right_expr)
+    if let (Some(literal), Some((inner_expr, _cast_type))) = (
+        binary.left().as_any().downcast_ref::<Literal>(),
+        extract_cast_info(binary.right()),
+    ) {
+        // For literal op cast(expr), we need to swap the operator
+        let swapped_op = swap_operator(*binary.op());
+        if let Some(unwrapped) = try_unwrap_cast_comparison(
+            Arc::clone(inner_expr),
+            literal.value(),
+            swapped_op,
+            schema,
+        )? {
+            // unwrapped is already the correct expression: inner_expr swapped_op casted_literal
+            return Ok(Some(unwrapped));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Extract cast information from a physical expression
```

Comment:
Can we please update this to explain what info is extracted? 

it seems like it looks for `CAST(expr, datatype)` and returns the `expr` and `datatype` if if finds a match